### PR TITLE
Rule: don't process if the incoming data is same

### PR DESCRIPTION
**Pull request by Aintaer**
Each rule will have to maintain its own cache of what affects it. Then
on an incoming process call, diff the cache to find out whether or not
it should process at all.

Additionally, when processing new data, the results may be different, so
the mark() operation must mark all affected nodes. When called to mark()
by DOM mutation, the results remain the same, so only mark the newly
affected nodes.

### CODE REVIEW

File: src/Rule.js [hunk 1]
```diff
@@ -31,6 +36,22 @@ define([
     },
 
     process: function(data, extensions) {
+      var artifacts = Object.keys(this.artifacts)
+      .reduce(function archaeology(artifacts, path) {
```
Comment (Attamusc) [inline src/Rule.js]: looooooooooooove this :P

File: src/Rule.js [hunk 2]
```diff
@@ -44,20 +65,26 @@ define([
...
 
-      this.mark();
+      this.mark(true);
     },
 
-    mark: function() {
-      var elements = css.find(this.computedSelector);
+    mark: function(fresh) {
+      var affected = css.find(this.computedSelector),
+      elements = Array.prototype.slice.call(affected);
```
Comment (joesepi) [inline src/Rule.js]: Let me just say that I hate when single var is used but indentation is not maintained. Whatever though.

**Comment by Attamusc** [comment]
:+1:

**Comment by joesepi** [comment]
:+1:

**Pull request merged**